src/Compilers/Core/Portable/SourceGeneration/SyntaxStore.cs

```diff
@@ -88,8 +88,23 @@ public IStateTable GetSyntaxInputTable(SyntaxInputNode syntaxInputNode, NodeStat
                             {
                                 try
                                 {
-                                    _cancellationToken.ThrowIfCancellationRequested();
-                                    syntaxInputBuilders[i].builder.VisitTree(root, state, model, _cancellationToken);
+                                    Stopwatch sw = Stopwatch.StartNew();
+                                    try
+                                    {
+                                        _cancellationToken.ThrowIfCancellationRequested();
+                                        syntaxInputBuilders[i].builder.VisitTree(root, state, model, _cancellationToken);
+                                    }
+                                    finally
+                                    {
+                                        var elapsed = sw.Elapsed;
+
+                                        // if this node isn't the one that caused the update, ensure we remember it and remove the time it took from the requester
+                                        if (syntaxInputBuilders[i].node != syntaxInputNode)
+                                        {
+                                            _syntaxTimes[syntaxInputNode] = _syntaxTimes[syntaxInputNode].Subtract(elapsed);
```

Comment:
>_syntaxTimes[syntaxInputNode].Subtract(elapsed) [](http://example.com/codeflow?start=76&length=47)

We're subtracting from `_syntaxTimes[syntaxInputNode]` but where did we add time for that node?

Reply:
It gets set to `TimeSpan.Zero` on line 73. This is the node that actually asked for the update, and so got unfairly attributed all the time of the other nodes. We remove the other nodes time from zero, so its offset is actually negative (the sum of all the time we recorded for it that was actually processing other nodes) so that it gets its time reduced and every other node gets its time increased.

Reply:
> so its offset is actually negative

Does that mean `_syntaxTimes[i]` is actually an offset rather than the elapsed time for node `i`? If so, where is the elapsed time calculated? (Sorry if I'm misunderstanding.)

Reply:
Yes, its an offset that gets applied to calculate the correct elapsed time. We do the actual timing in the generator driver here: https://github.com/dotnet/roslyn/blob/ab190e4c1efb9115da60ecdf27becea40740dbc9/src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs#L261

That sits around the entire graph calculation operation for a given generator, but because we do the syntax nodes in parallel whenever the first one is requested, we have to adjust via the offset to ensure we actually have the correct times for each generator, and we don't unfairly claim that one generator was responsible for all the time taken to calculate syntax nodes, and the others took no time.